h/settings.py

```diff
@@ -13,6 +13,52 @@ class SettingError(Exception):
     pass
 
 
+class SettingsManager(object):
+    def __init__(self, settings, environ):
+        self.settings = settings
+        self.environ = environ
+
+    def set(self,
+            setting,
+            env_var,
+            type=str,
+            required=False,
+            default=None,
+            deprecated_msg=None):
+        """
+        Sets the specified setting in settings to the value of the env_var,
+        of if the env_var isn't present and a default value is specified; to
+        the default value..
+        If the env_var doesn't exist and the setting is not required then it
+        does nothing.
+        If the setting is required and not specified it raises a SettingError.
+        If a default value is provided and the env_var doesn't exist, the setting will be
+        overriden with the provided default value.
+        """
```

Comment:
I realise this is draft code so this  probably wasn't intended as a final docstring, but would you mind having a go at rewriting this docstring to follow [PEP 257](https://www.python.org/dev/peps/pep-0257/)? I think it might be clearer. Some particular details from PEP 257:

* First line of docstring should be a concise summary that fits on one line, followed by a blank line and then the more detailed description.

  (Tip: I think it's okay to just say something like `Set ``setting`` in the Pyramid settings` for the first line, details like how an envvar or default might be used, or the setting might be left unchanged, can go in the rest of the docstring.)

* The first line should "prescribe the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...". So in your example it would be "Set the setting in settings" rather than "Set**s** the setting in settings".

  This may seem nit-picky but generally imperative tense leads to shorter and clearer sentences, my example docstring at the bottom of this comment might make this clearer. Another way of putting this is: write your docstrings (and comments, commit messages, pull request descriptions, ...) like pseudo-code. (We also use this in [our git commit style](https://chris.beams.io/posts/git-commit/) btw.)

* Since this method has a lot of parameters I think it would be worth documenting each of them individually using `:param foo: ...` and `:type foo: ...` lines in the docstring. These `:param` and `:type` etc things are Sphinx "info field lists", [docs here](http://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists)

* Since this method also raises a certain exception under certain condition I think it'd be worth documenting that with `:raises SettingError: ...` too (see info field list docs link above)

General docstring writing tip: break it up into small pieces! It will make things clearer to separate different cases and details into different paragraphs. You've started to do this after the first long line of your docstring, but haven't left blank lines between the lines. For example `"Sets the specified setting in settings to the value of the env_var, or if the env_var isn't present and a default value is specified; to the default value.."` is probably clearer broken up case-by-case something like this:

> Set `settings[name]` to `envvar` or `default`.
> 
> Environment variables override config file settings and defaults: if the named `envvar` is in the environment then set `settings[name]` to the value of the `envvar`.
> 
> Config file settings override defauits: if the `envvar` isn't in the environment, and the named setting is already present in settings, then leave `settings[name]` unchanged.
> 
> Defaults are used if the setting isn't in the environment or the config file: if  `envvar` isn't in the environment and `name` isn't in settings then set `settings[name]` to the given `default` value.
> 
> If  `envvar` isn't in the environment and `name` isn't in settings and no `default` is given then leave `settings[name]` unset. If this happens and `required=True` then raise `SettingError`.

(Note that I'm also using the actual names of the method params in the docstring, and even using `settings[name]`, I think this makes it much clearer than saying indirect things like "the specified setting".)